app/models/detector/citation.rb

```diff
@@ -0,0 +1,156 @@
+    end
+
+    private
+
+    # This combines the two reports generated by the Citation detector (subpatterns and summary), and calculates the
+    # final score value from their contents.
+    #
+    # Any detected subpattern is counted toward the score (multiple detections do not get counted twice). For example,
+    # if the brackets pattern finds two matches, it still only adds one to the final score.
+    #
+    # For the summary report, each value is compared with a threshold value in the SUMMARY_THRESHOLDS constant. The
+    # number of values which meet or exceed their threshold are added to the score. As an example, if a search term has
+    # five words, this value is compared to the word threshold (also five). Because the threshold is met, the score gets
+    # incremented by one.
+    #
+    # @return integer
+    def calculate_score
+      summary_score = @summary.count do |key, value|
+        SUMMARY_THRESHOLDS.key?(key) && value >= SUMMARY_THRESHOLDS[key]
+      end
+
+      summary_score + @subpatterns.length
+    end
+
+    # This calculates the number of characters in the search term. It is called by the summarize method.
+    def characters(term)
+      term.length
+    end
+
+    # This counts the number of colons that appear in the search term, because they tend to appear more often in
+    # citations than in other searches. It is called by the summarize method.
+    def colons(term)
+      term.count(':')
+    end
+
+    # This counts the number of commas in the search term. It is called by the summarize method.
+    def commas(term)
+      term.count(',')
+    end
+
+    # This builds one of the two main components of the Citation detector - the subpattern report. It uses each of the
+    # regular expressions in the CITATION_PATTERNS constant, extracting all matches using the scan method.
+    #
+    # @return hash
+    def pattern_checker(term)
+      CITATION_PATTERNS.each_pair do |type, pattern|
+        @subpatterns[type.to_sym] = scan(pattern, term) if scan(pattern, term).present?
+      end
+    end
+
+    # This counts the number of periods in the search term. It is called by the summarize method.
+    def periods(term)
+      term.count('.')
+    end
+
+    # This is a convenience method for the scan method, which is used by pattern_checker.
+    def scan(pattern, term)
+      term.scan(pattern).map(&:strip)
+    end
+
+    # This counts the semicolons in the search term. It is called by the summarize method.
+    def semicolons(term)
+      term.count(';')
+    end
+
+    # This builds one of the two main components of the Citation detector - the summary report. It calls each of the
+    # methods in the first line - which all return integers - and puts the result as a key-value pair in the @summary
+    # instance variable.
+    #
+    # @return hash
+    def summarize(term)
+      %w[characters colons commas periods semicolons words].each do |check|
+        @summary[check.to_sym] = send(check, term)
```

Comment:
This took me a bit to follow what was going on even with the docs. It's a solid pattern, but we don't reach for it enough for me to have immediately understood it.

Reply:
Is there a better way to say "call all these methods" ? I was looking for a structure that was efficient without being opaque, so I'm open to hearing that this doesn't quite hit that spot.

Reply:
No, I think what you are doing here is great. `send` is meant for this. I just had to look it up as we don't use it much.